crates/astria-conductor/src/sequencer/mod.rs

```diff
@@ -0,0 +1,252 @@
+        );
+
+        let mut sync = sync::run(
+            start_sync_height,
+            latest_height,
+            pool.clone(),
+            executor_tx.clone(),
+        )
+        .boxed()
+        .fuse();
+
+        let mut sync_done = Some(sync_done);
+        let mut resubscribe = future::Fuse::terminated();
+        'reader_loop: loop {
+            select! {
+                shutdown = &mut shutdown => {
+                    let ret = match shutdown {
+                        Err(e) => {
+                            warn!(error.message = %e, "shutdown channel closed unexpectedly; shutting down");
+                            Err(e).wrap_err("shut down channel closed unexpectedly")
+                        }
+                        Ok(()) => {
+                            info!("received shutdown signal; shutting down");
+                            Ok(())
+                        }
+                    };
+                    break 'reader_loop ret;
+                }
+
+                res = &mut sync, if !sync.is_terminated() => {
+                    if let Err(e) = res {
+                        warn!(error.message = %e, error.cause = ?e, "sync failed; continuing with normal operation");
+                    } else {
+                        info!("sync finished successfully");
+                    }
+                    let sync_done = sync_done.take().expect("channel should only be used once and only in this branch; this is a bug");
+                    let _ = sync_done.send(());
+                }
+
+                // New blocks from the subscription to the sequencer. If this fused stream ever returns `None`,
+                // a task is scheduled to resubscribe to new blocks.
+                // Blocks are pushed into `pending_blocks` so that they are forwarded to the executor in the
+                // order they were received.
+                new_block = new_blocks.next(), if !new_blocks.is_done() => {
+                    schedule_for_forwarding_or_resubscribe(
+                        &mut resubscribe,
+                        &mut pending_blocks,
+                        pool.clone(),
+                        new_block,
+                    );
+                }
+
+                // Regular pending blocks will be submitted to the executor in the order they were received.
+                // The condition on `sync` ensures that blocks from the sync process are forwarded first.
+                Some(res) = pending_blocks.next(), if sync.is_terminated() => {
+                    if let Err(e) = forward_pending_block(executor_tx.clone(), res) {
+                        error!("failed forwarding blocks during regular operation; exiting reader");
+                        break 'reader_loop Err(e).wrap_err("failed forwarding blocks regular operation");
+                    }
+                }
+
+                res = &mut resubscribe => {
```

Comment:
what is `res` on the first iteration where this is called? since `let mut resubscribe = future::Fuse::terminated();?

Reply:
`res` would be nothing, since this future will never return anything (if polled, it will yield `Pending` but never `Ready`). I have just added `, if !resubscribe.is_terminated()` to the select arm however to make this clearer.